File: week-1/lecture/gcd.py
"""Gcd example.

This file belongs to the introductory lecture examples on recursion, searching,
and simple sorting. It uses Euclid's algorithm to compute the greatest common
divisor of two integers.
"""


def gcd(m, n):
    """Compute the greatest common divisor using recursive Euclid's algorithm.

    Args:
        m (int): The m input used by this function.
        n (int): The n input used by this function.

    Returns:
        int: The integer result produced by the algorithm.
    """
    m = abs(m)
    n = abs(n)

    # Euclid's algorithm still works if we swap the values so the larger one
    # comes first.
    if m < n:
        m, n = n, m

    if n == 0:
        return m

    return gcd(n, m % n)


def main():
    """Run the small demonstration for this teaching example.

    Returns:
        None: This function updates data or prints results and does not return a value.
    """
    m = 60
    n = 24
    print(f"gcd({m}, {n}) = {gcd(m, n)}")


if __name__ == "__main__":
    main()

File: week-1/lecture/gcd.cpp
/**
 * @file gcd.cpp
 * @brief C++ version of the recursive Euclid algorithm example.
 */

#include <cstdlib>
#include <iostream>

using namespace std;

/**
 * @brief Compute the greatest common divisor of two integers.
 * @param m int First integer. The function treats negative values by using their absolute value.
 * @param n int Second integer. The function repeatedly replaces the larger value with the remainder.
 * @return int Greatest common divisor of m and n.
 */
int gcd(int m, int n) {
    m = abs(m);
    n = abs(n);

    if (m < n) {
        swap(m, n);
    }
    if (n == 0) {
        return m;
    }
    return gcd(n, m % n);
}

/**
 * @brief Run a small demonstration of the gcd function.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    int m = 60;
    int n = 24;
    cout << "gcd(" << m << ", " << n << ") = " << gcd(m, n) << '\n';
    return 0;
}
